Add more to data loader guide (#2502)
graphql-dotnet/graphql-dotnet: docs2/site/docs/guides/dataloader.md

        IDictionary<int, Order> data = await _dbContext.Orders.Where(order => ids.Contains(order.Id)).ToDictionaryAsync(x => x.Id, cancellationToken);
        foreach (DataLoaderPair<int, Order> entry in list)
        {
            entry.SetResult(data.TryGetValue(entry.Key, out var order) ? order : null);
        }
    }
}

// similar to CollectionBatchDataLoader
public class MyOrderItemsDataLoader : DataLoaderBase<int, IEnumerable<OrderItem>>
{
    private readonly MyDbContext _dbContext;
    public MyOrderItemsDataLoader(MyDbContext dataContext)
    {
        _dbContext = dataContext;
    }

    protected override Task FetchAsync(IEnumerable<DataLoaderPair<int, IEnumerable<OrderItem>>> list, CancellationToken cancellationToken)
    {
        IEnumerable<int> ids = list.Select(pair => pair.Key);
        IEnumerable<OrderItem> data = await _dbContext.OrderItems.Where(orderItem => ids.Contains(orderItem.OrderId)).ToListAsync(cancellationToken);
        ILookup<int, OrderItem> dataLookup = data.ToLookup(x => x.OrderId);
        foreach (DataLoaderPair<int, IEnumerable<OrderItem>> entry in list)
        {
            entry.SetResult(dataLookup[entry.Key]);
        }
    }
}
```

> You will need to regsiter the data loaders as a scoped service within your DI framework.

Reviewer: ```suggestion
You will need to register the data loader as a scoped service within your DI framework.
```

```csharp
services.AddScoped<MyOrderDataLoader>();

> services.AddScoped<MyOrderItemDataLoader>();

Reviewer: ```suggestion
services.AddScoped<MyOrderItemsDataLoader>();
```

```

Then within your field resolvers, access the data loader via the `RequestServices` property and call `LoadAsync` as before:

```csharp
public class MyQuery : ObjectGraphType
{
    public MyQuery()
    {
        Field<OrderType, Order>()
            .Name("Order")
            .Argument<IdGraphType>("id")
            .ResolveAsync(context =>
            {
                // Get the custom data loader
                var loader = context.RequestServices.GetRequiredService<MyOrderDataLoader>();

>                 // Add this UserId to the pending keys to fetch

Reviewer: ```suggestion
                // Add this UserId to the pending keys to fetch.
```

                // The execution strategy will trigger the data loader to fetch the data via MyOrderDataLoader.FetchAsync() at the

>                 //   appropriate time, and the field will be resolved with an instance of Order once FetchAsync()

Reviewer: ```suggestion
                // appropriate time, and the field will be resolved with an instance of Order once FetchAsync()
```

>                 //   returns with the batched results

Reviewer: ```suggestion
                // returns with the batched results
```

    {
        using (var scope = _rootServiceProvider.CreateScope())
        {
            MyDbContext dbContext = scope.ServiceProvider.GetRequiredService<MyDbContext>();
            IEnumerable<int> ids = list.Select(pair => pair.Key);
            IDictionary<int, Order> data = await dbContext.Orders.Where(order => ids.Contains(order.Id)).ToDictionaryAsync(x => x.Id, cancellationToken);
            foreach (DataLoaderPair<int, Order> entry in list)
            {
                entry.SetResult(data.TryGetValue(entry.Key, out var order) ? order : null);
            }
        }
    }
}
```

As a singleton, you can pull the singleton instance into your graphtype class in its constructor.

```csharp
    public class MyQuery : ObjectGraphType
    {
        public MyQuery(MyOrderDataLoader loader)
        {
            Field<OrderType, Order>()
                .Name("Order")
                .Argument<IdGraphType>("id")
                .ResolveAsync(context =>
                {
                    return loader.LoadAsync(context.GetArgument<int>("id"));
                });
        }
    }
```

## Adding a global cache

> Data loaders will, by default, cache values returned for a given key for the lifetime of a request. You can change the

Reviewer: ```suggestion
Data loaders will, by default, cache values returned for a given key for the lifetime of a request. You can change
```

```csharp
public class MyOrderDataLoader : DataLoaderBase<int, Order>
{
    private readonly IServiceProvider _rootServiceProvider;
    private readonly IMemoryCache _memoryCache;
    private readonly MemoryCacheEntryOptions _memoryCacheEntryOptions;
    private const string CACHE_PREFIX = "ORDER_";
        
    public MyOrderDataLoader(IServiceProvider serviceProvider, IMemoryCache memoryCache) : base(false)
    {
        _rootServiceProvider = serviceProvider;
        _memoryCache = memoryCache;
        _memoryCacheEntryOptions = new MemoryCacheEntryOptions
        {
            // specify a maximum lifetime of 5 minutes
            AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(5),
            // set so that the size of the cache can be limited
            Size = 1,
        };
    }

    protected override async Task FetchAsync(IEnumerable<DataLoaderPair<int, Order>> list, CancellationToken cancellationToken)
    {
        // create a list of keys that are not in the cache
        var unMatched = new List<DataLoaderPair<int, Order>>(list.Count());
        // attempt to match any keys possible from the global cache

>         foreach(var entry in list)

Reviewer: ```suggestion
        foreach (var entry in list)
```
